crates/aptos-drop-helper/src/lib.rs

```diff
@@ -0,0 +1,108 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+mod metrics;
+
+use crate::metrics::TIMER;
+use aptos_experimental_runtimes::thread_manager::THREAD_MANAGER;
+use aptos_infallible::Mutex;
+use aptos_metrics_core::TimerHelper;
+use once_cell::sync::Lazy;
+use std::sync::mpsc::{channel, Receiver, Sender};
+
+pub static DEFAULT_DROP_HELPER: Lazy<DropHelper> = Lazy::new(|| DropHelper::new("default", 128));
+
+/// A helper to send things to a thread pool for asynchronous dropping.
+///
+/// Be aware that there is a bounded number of concurrent drops, as a result:
+///   1. when it's "out of capacity", `schedule_drop` will block until a slot to be available.
+///   2. if the `Drop` implementation tries to lock things, there can be a potential dead lock due
+///      to another thing being waiting for a slot to be available.
+pub struct DropHelper {
+    name: &'static str,
+    token_tx: Sender<()>,
+    token_rx: Mutex<Receiver<()>>,
+}
+
+impl DropHelper {
+    pub fn new(name: &'static str, max_concurrent_drops: usize) -> Self {
+        let (token_tx, token_rx) = channel();
+        for _ in 0..max_concurrent_drops {
+            token_tx.send(()).expect("Failed to buffer initial tokens.");
+        }
+        Self {
+            name,
+            token_tx,
+            token_rx: Mutex::new(token_rx),
+        }
+    }
+
+    pub fn schedule_drop<V: Send + 'static>(&self, v: V) {
+        self.schedule_drop_impl(v, None)
+    }
+
+    pub fn schedule_drop_with_waiter<V: Send + 'static>(&self, v: V) -> Receiver<()> {
+        let (tx, rx) = channel();
+        self.schedule_drop_impl(v, Some(tx));
+        rx
+    }
+
+    fn schedule_drop_impl<V: Send + 'static>(&self, v: V, notif_sender_opt: Option<Sender<()>>) {
+        let _timer = TIMER.timer_with(&[self.name, "enqueue_drop"]);
+
+        self.token_rx.lock().recv().unwrap();
```

Comment:
if we don't lock, what problem could happen here?

Reply:
You can't receive from multiple threads at the same time for a mpsc channel. (can't compile either)